### DISCUSSION TITLE
strange if statment on root, and proper text sizing

### DISCUSSION THREAD

**Discussion by moolight-seashell** [slint-ui/slint#5175] [Q&A]
By default, it seems that a text element defines the size of its parent.

For example, if I create a gridlayout with rectangle, everything is the same size.
If I put text inside one of the rectangle, the whole cell takes the size of the text.

It's actually quite annoying that adding text to a sub-element ruins the formatting of all the parent elements.

I try to make the size of the text adapt to the size of the element.
I've found a trick to do this, but it seems quite dirty. Is this a bug?

the non working code : 
![image](https://github.com/slint-ui/slint/assets/49676861/31d340bc-11b8-4faa-a5ad-d4dbaabfebfc)

and an if solve this, why ??
![image](https://github.com/slint-ui/slint/assets/49676861/2e5f835d-c208-457e-85e3-ec1472c6808c)

Oh wait what ? this also works with an if(true) 😅

Could it be that the "if" is annalyzed at runtime and that slint doesn't take the element into account for the size calculation, or something like that ? And how to do this without the very dirty if(true) ?

**Comment by moolight-seashell** [comment]
there is a lot of missing option here but this works quite well even with gridlayout , 

```slint
component AdaptiveText {
    in property <color> background;
    in property <color> color;
    in property <string> text;
    in property <int> text_length;

    property <length> with_size: root.width/(text_length/2);
    property <length> height_size: root.height;
    property <length> true_size;

    states [
        active when (with_size > height_size):{
            true_size: height_size;
        }
        active when (with_size < height_size):{
            true_size: with_size;
        }
    ]

    Rectangle {
        if(true): Text{ // ?? break/skipp static size analysis, dirty but allow to use root based size 
            // option ? if(true_size != 0px)
            text: text;
            color: color;
            font-size: true_size;
            horizontal-alignment: center;
            vertical-alignment: center;
        }
        background: background;
    }
}
component TestAdText {
    GridLayout {
        AdaptiveText {background: black;text: "testtas";color: white; text_length: 7;}
    }
}
```

is there a way to get the amount of chars like a .len() ? and a function to get the greatest amount of char for a multiple line text , and maybe the amount of lines ?

**Comment by ogoffart** [comment]
Yes, there is a difference for the computation of the layout constraint of the parent when the element is in a `if`. This is a current limitation that is tracked by issue #407 

We can get around that by specifying the layout constraint (minimum / preferred size) manually.
This is not ideal, i know.

**Comment by moolight-seashell** [comment]
ok great, thanks for the info.
It seems that in the case of the issue https://github.com/slint-ui/slint/issues/407 this if is a problem, whereas in my case it's actually a big advantage
but it doesn't tell me if there's a better way to make an adaptive text sizing without a bind loop
and if is there a length() function for string properties ?
